### PR TITLE
Lightstep

### PR DESCRIPTION
**Pull request by kardolus**
Not super happy about the `WithTracer` implementation: [here](https://github.com/sidelight-labs/libhttp/blob/60e0bbd2d8c84dd6ca6f07198fdf7bf1fb151448/client/callout.go#L80). I may not be understanding the patern.

This PR adds the option for tracing to the http-library. May move some of this to `libc` if this is reusable for db calls.

Added `./scripts/unit.sh`.

### CODE REVIEW

File: client/callout.go [hunk 1]
```diff
@@ -63,13 +69,18 @@ func (c *Callout) buildRequestWithOptions(method string, url string, reqBody str
...
 	}
 
 	for _, option := range options {
 		option(requestOpts)
 	}
 
+	if requestOpts.tracer != nil {
+		c.defaultTracer = requestOpts.tracer
+	}
+
```
Comment (aemengo) [inline client/callout.go:83]: This shouldn't be necessary. The `for` loop above should take care of things.

File: tracing/lightstep_tracer.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	var span trace.Span
+
+	if t.spanID == "" {
+		t.ctx, span = t.tracer.Start(t.ctx, name)
+		t.spanID = span.SpanContext().SpanID().String()
+	} else {
+		_, span = t.tracer.Start(t.ctx, name)
+	}
```
Comment (aemengo) [inline tracing/lightstep_tracer.go:48]: I'm having trouble seeing the motivation behind this. Are we trying to protect the developer from calling `Trace(name string) Span` twice?

File: tracing/lightstep_tracer.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+type LightstepSpan struct {
+	span trace.Span
+}
+
+func (s *LightstepSpan) End() {
+	if s.span != nil {
+		s.span.End()
+	}
+}
```
Comment (aemengo) [inline tracing/lightstep_tracer.go:65]: I'm having trouble seeing what this abstraction gives us. If `trace.Span` is returned directly, it won't ever  be `nil`.

**Review by aemengo — CHANGES_REQUESTED** [review]
I tried

* `client.WithDefaultTracer(tracing.NewLightstepTracer(service, token))`
* `client.WithTracer(tracing.NewLightstepTracer(service, token))`

 They both weren't working for me. Can you confirm?

**Comment by kardolus** [comment]
> I tried
> 
> * `client.WithDefaultTracer(tracing.NewLightstepTracer(service, token))`
> * `client.WithTracer(tracing.NewLightstepTracer(service, token))`
> 
> They both weren't working for me. Can you confirm?

This code works for me: https://github.com/sidelight-labs/osmosis/blob/e0a7dc6e2cf934a4036ad44550d1d9bfabd44844/commands/parse-block.go#L55

File: tracing/lightstep_tracer.go (hunk 2 above)
Comment (kardolus) [reply tracing/lightstep_tracer.go:48]: This is a hack. `t.ctx` must only be overwritten if there is no `spanID`. If you overwrite the context while it has a  `spanID` the next span appears as a child of the previous span. So if you do that on every iteration you will see something like a linkedlist of spans. We want more of a trie of spans where the first span is the parent and all other spans are direct children of the root.

File: client/callout.go (hunk 1 above)
Comment (kardolus) [reply client/callout.go:83]: Hmm, my test failed without this logic. I must doing something wrong in the for-loop then.

Comment (kardolus) [reply client/callout.go:83]: This is the test that will fail without the if-statement: 
```
overwrites the default setting
```

File: tracing/lightstep_tracer.go (hunk 3 above)
Comment (kardolus) [reply tracing/lightstep_tracer.go:65]: You mean we should ditch the if-statement? Or is this a question around the existence of the interface?

File: tracing/lightstep_tracer.go (hunk 2 above)
Comment (kardolus) [reply tracing/lightstep_tracer.go:48]: I'm not sure how to solve this issue without setting the context on the struct. Through the context the span knows which parent trace it has. If we have multiple http calls that are part of a parent process (such as parsing a block), we need to pass that parent trace along to all the spans. 

Alternatively, we could recreate a context for every http call. So we store individual context values on the struct and make a new context based on those values.

**Comment by mdelillo** [comment]
I've been reading up more about OpenTelemetry ([this doc](https://opentelemetry.lightstep.com/spans/) was particularly useful), and I don't think we want to put too much of this logic into `libhttp`.

Creating a new tracer that can only create spans in a single way is very limiting. It ties us to one specific way of nesting spans and makes it difficult to have a span that includes things other than http calls (such as db requests). So ideally we're creating the tracers outside of this library, which leaves creating/ending spans. But that is also limiting - it means we can't name them differently, can't attach attributes (such as success/failure), etc.

I'd propose removing the `tracing` package and just creating the tracer directly in the top-level of our apps. For a very basic request trace we could have a `WithTrace` request option that takes a `Tracer` and `Context` and creates a `Span` for the request, similar to what this PR does currently. It would ignore the created context, so each request would be its own span (nested under the passed context if that already is a span). Maybe also a client option that does the same by default. Usage would look like:

```go
// set up Launcher

tracer := otel.Tracer("my-service")

ctx, span := tracer.Start(context.Background(), "parent 1")
callout := client.New()
callout.Get("some-url", WithTrace(tracer, ctx))
callout.Get("some-other-url", WithTrace(tracer, ctx))
span.End()

ctx, span = tracer.Start(context.Background(), "parent 2")
tracingCallout := client.New(WithDefaultTrace(tracer, ctx))
tracingCallout.Get("some-url")
tracingCallout.Get("some-other-url")
span.End()
```

Both of those create a parent span with two nested spans, and the parent can continue to be used elsewhere.

File: tracing/lightstep_tracer.go (hunk 3 above)
Comment (aemengo) [reply tracing/lightstep_tracer.go:65]: The existence of the interface. Is the here for testing?

**Comment by kardolus** [comment]
I like the idea of passing in a context a lot. Gives you way more control; for example to set a custom trace-name. It's kinda close to my original implementation with `WithContext` and `WithTracing`. But this solution with just one method is more elegant. I will try to find some time to rewrite.

**Comment by kardolus** [comment]
@mdelillo @aemengo Sorry y'all! I'm totally going in circles on this PR. Now I am back at using open telemetry instead of playing with our own Tracer interface. But now I do not have unit tests for the open-telemetry tracer.

**Review by aemengo — APPROVED** [review]
Working for me. Thanks for showing me how to use.

_Suggestions:_
* Since we have retry logic, it might be helpful to put attempt in the `spanName`. e.g. `GET / (attempt 1/3)`
* Prefer `requestOpts` over `defaultOpts` to limit possible code paths
```diff
diff --git a/client/callout.go b/client/callout.go
index 31b0d3a..6b1ec49 100644
--- a/client/callout.go
+++ b/client/callout.go
@@ -82,16 +82,13 @@ func (c *Callout) buildRequestWithOptions(method string, url string, reqBody str
 		headers: c.defaultHeaders,
 		retries: c.defaultRetries,
 		tracer:  c.defaultTracer,
+		context: c.defaultContext,
 	}
 
 	for _, option := range options {
 		option(requestOpts)
 	}
 
-	if requestOpts.tracer != nil {
-		c.defaultTracer = requestOpts.tracer
-	}
-
 	var reqBodyReader io.Reader
 	if reqBody != "" {
 		reqBodyReader = strings.NewReader(reqBody)
@@ -111,7 +108,7 @@ func (c *Callout) buildRequestWithOptions(method string, url string, reqBody str
 	var statusCode int
 	var body []byte
 	for i := 0; i <= requestOpts.retries; i++ {
-		body, statusCode, err = c.doRequest(req, requestOpts.bodyWriter)
+		body, statusCode, err = c.doRequest(req, requestOpts.bodyWriter, requestOpts)
 		if err != nil {
 			return nil, err
 		}
@@ -137,9 +134,9 @@ func (c *Callout) buildRequestWithOptions(method string, url string, reqBody str
 	}
 }
 
-func (c *Callout) doRequest(req *http.Request, writer io.Writer) ([]byte, int, error) {
-	if c.defaultTracer != nil {
-		_, span := c.defaultTracer.Start(c.defaultContext, req.URL.Path)
+func (c *Callout) doRequest(req *http.Request, writer io.Writer, opts *requestOptions) ([]byte, int, error) {
+	if opts.tracer != nil {
+		_, span := opts.tracer.Start(opts.context, req.URL.Path)
 		defer span.End()
 	}
 
```

**Comment by kardolus** [comment]
@aemengo dude I love that refactor. That's exactly what I couldn't come up with. I hated that if-statement I had. This is way nicer. Just made the change.

Love the idea of the retries too. Will attempt to do that in the next commit.

**Pull request merged**